Reply/app/src/main/java/com/example/reply/ui/navigation/ReplyNavigationComponents.kt

```diff
@@ -52,33 +54,43 @@ import androidx.compose.ui.res.stringResource
 import androidx.compose.ui.text.style.TextAlign
 import androidx.compose.ui.unit.dp
 import com.example.reply.R
+import com.example.reply.ui.utils.ReplyNavigationContentPosition
 
 
 @Composable
 fun ReplyNavigationRail(
     selectedDestination: String,
+    navigationContentPosition: ReplyNavigationContentPosition,
     navigateToTopLevelDestination: (ReplyTopLevelDestination) -> Unit,
     onDrawerClicked: () -> Unit = {},
 ) {
-    NavigationRail(modifier = Modifier.fillMaxHeight()) {
-        NavigationRailItem(
-            selected = false,
-            onClick = onDrawerClicked,
-            icon =  { Icon(imageVector = Icons.Default.Menu, contentDescription = stringResource(id = R.string.navigation_drawer)) }
-        )
-        FloatingActionButton(
-            onClick =  { /*TODO*/ },
-            modifier = Modifier.padding(top = 18.dp, bottom = 40.dp),
-            containerColor = MaterialTheme.colorScheme.tertiaryContainer,
-            contentColor = MaterialTheme.colorScheme.onTertiaryContainer
-        ) {
-            Icon(
-                imageVector = Icons.Default.Edit,
-                contentDescription = stringResource(id = R.string.edit),
-                modifier = Modifier.size(18.dp)
+    NavigationRail(
+        modifier = Modifier.fillMaxHeight(),
+        header = {
+            NavigationRailItem(
+                selected = false,
+                onClick = onDrawerClicked,
+                icon =  { Icon(imageVector = Icons.Default.Menu, contentDescription = stringResource(id = R.string.navigation_drawer)) }
             )
+            FloatingActionButton(
+                onClick =  { /*TODO*/ },
+                modifier = Modifier.padding(top = 8.dp, bottom = 32.dp),
+                containerColor = MaterialTheme.colorScheme.tertiaryContainer,
+                contentColor = MaterialTheme.colorScheme.onTertiaryContainer
+            ) {
+                Icon(
+                    imageVector = Icons.Default.Edit,
+                    contentDescription = stringResource(id = R.string.edit),
+                    modifier = Modifier.size(18.dp)
+                )
+            }
+        }
+    ) {
+        //TODO ticket to support nav rail positioning out of box according to specs: b/232495216
+        if  (navigationContentPosition == ReplyNavigationContentPosition.BOTTOM
+            || navigationContentPosition == ReplyNavigationContentPosition.CENTER) {
+                Spacer(modifier = Modifier.weight(0.4f))
```

Comment:
This `0.4f` weight approximates the resulting centering with the header, right? Could we do that calculation manually to have it be perfectly centered?

In Compose doing like that should be fairly simple:

- Compute the size of the header
- Place the header at the top
- Compute the size of the rest
- Try placing it where we want it to be (based on the navigationContentPosition), but coerced to the bottom of the header region

Reply:
The issue is NavigationRail doesn't expose ability to place the header at top, Also column scope modifier is not exposed to position. May be you can share any code snippet if you have on above.

Reply:
Here's the sort of direct calculation we could do to center the items, if the header space allows it:

```
NavigationRail(
        modifier = Modifier.fillMaxHeight(),
    ) {
        Layout(
            content = {
                Column(
                    modifier = Modifier.layoutId("header"),
                    horizontalAlignment = Alignment.CenterHorizontally,
                    verticalArrangement = Arrangement.spacedBy(4.dp) // NavigationRailVerticalPadding
                ) {
                    NavigationRailItem(
                        selected = false,
                        onClick = onDrawerClicked,
                        icon =  { Icon(imageVector = Icons.Default.Menu, contentDescription = stringResource(id = R.string.navigation_drawer)) }
                    )
                    FloatingActionButton(
                        onClick =  { /*TODO*/ },
                        modifier = Modifier.padding(top = 8.dp, bottom = 32.dp),
                        containerColor = MaterialTheme.colorScheme.tertiaryContainer,
                        contentColor = MaterialTheme.colorScheme.onTertiaryContainer
                    ) {
                        Icon(
                            imageVector = Icons.Default.Edit,
                            contentDescription = stringResource(id = R.string.edit),
                            modifier = Modifier.size(18.dp)
                        )
                    }
                    Spacer(Modifier.height(8.dp)) // NavigationRailHeaderPadding
                    Spacer(Modifier.height(4.dp)) // NavigationRailVerticalPadding
                }

                Column(
                    modifier = Modifier.layoutId("content"),
                    horizontalAlignment = Alignment.CenterHorizontally,
                    verticalArrangement = Arrangement.spacedBy(4.dp) // NavigationRailVerticalPadding
                ) {
                    TOP_LEVEL_DESTINATIONS.forEach { replyDestination ->
                        NavigationRailItem(
                            selected = selectedDestination == replyDestination.route,
                            onClick = { navigateToTopLevelDestination.invoke(replyDestination) },
                            icon = {
                                Icon(
                                    imageVector = replyDestination.selectedIcon,
                                    contentDescription = stringResource(id = replyDestination.iconTextId)
                                )
                            }
                        )
                    }
                }
            },
            measurePolicy = { measurables, constraints ->
                lateinit var headerMeasurable: Measurable
                lateinit var contentMeasurable: Measurable
                measurables.forEach {
                    when (it.layoutId) {
                        "header" -> headerMeasurable = it
                        "content" -> contentMeasurable = it
                        else -> error("Unknown layoutId encountered!")
                    }
                }

                val headerPlaceable = headerMeasurable.measure(constraints)
                val contentPlaceable = contentMeasurable.measure(
                    constraints.offset(vertical = -headerPlaceable.height)
                )
                layout(constraints.maxWidth, constraints.maxHeight) {
                    // Place the header, this goes at the top
                    headerPlaceable.placeRelative(0, 0)

                    // Determine how much space is not taken up by the content
                    val nonContentVerticalSpace = constraints.maxHeight - contentPlaceable.height

                    val contentPlaceableY = when (navigationContentPosition) {
                        // Figure out the place we want to place the content, with respect to the
                        // parent (ignoring the header for now)
                        ReplyNavigationContentPosition.TOP -> 0
                        ReplyNavigationContentPosition.CENTER -> nonContentVerticalSpace / 2
                        ReplyNavigationContentPosition.BOTTOM -> nonContentVerticalSpace
                    }
                        // And finally, make sure we don't overlap with the header.
                        .coerceAtLeast(headerPlaceable.height)

                    contentPlaceable.placeRelative(0, contentPlaceableY)
                }
            }
        )
    }
```

Reply:
Thanks for this suggestion, Really helpful and have added it.